File: complete/src/main/java/hello/GreetingController.java
package hello;

import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PostMapping;

@Controller
public class GreetingController {

    @GetMapping("/greeting")
    public String greetingForm(Model model) {
        model.addAttribute("greeting_attribute", new GreetingClass());
        return "greeting_template";
    }

    @PostMapping("/greeting")
    public String greetingSubmit(@ModelAttribute GreetingClass greetingClass) {
        return "result";
    }
}

File: complete-kotlin/src/main/kotlin/com/example/handlingformsubmission/GreetingController.kt
package com.example.handlingformsubmission

import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ModelAttribute
import org.springframework.web.bind.annotation.PostMapping

@Controller
class GreetingController {

    @GetMapping("/greeting")
    fun greetingForm(model: Model): String {
        model.addAttribute("greeting", Greeting())
        return "greeting"
    }

    @PostMapping("/greeting")
    fun greetingSubmit(@ModelAttribute greeting: Greeting, model: Model): String {
        model.addAttribute("greeting", greeting)
        return "result"
    }
}
